include/chainbase/chainrocks.hpp

```diff
@@ -0,0 +1,822 @@
+
+      index(const index&)= delete;
+      index& operator= (const index&) = delete;
+
+      index(index&&)= delete;
+      index& operator= (index&&) = delete;
+
+      //////////////////////////////////
+      /// Temporary helper; remove later
+      void print_state() {
+         std::cout << "_state:\n";
+         for (const auto& value : _state) {
+            std::cout << "value.first: " << value.first << ' ' << "value.second: " << value.second << '\n';
+            // std::cout << value.first << value.second << ' ';
+         }
+         std::cout << '\n';
+      }
+
+      //////////////////////////////////
+      /// Temporary helper; remove later
+      void print_keys() {
+         std::cout << "print_keys()\n";
+         for (const auto& undo_state_obj : _stack) {
+            std::cout << "_new_keys: ";
+            for (const auto& key : undo_state_obj._new_keys) {
+               std::cout << key << ' ';
+            }
+            std::cout << '\n';
+         }
+      }
+
+      const auto& state() const {
+         return _state;
+      }
+
+      const auto& stack() const {
+         return _stack;
+      }
+
+      /// Add a new value to `_state` or modifies an existing value.
+      void put(const uint64_t& key, const std::string& value) {
+         _on_put(key, value);
+         _state[key] = value;
+      }
+
+      /// Remove a value from `_state`.
+      void remove(const uint64_t& key) {
+         _on_remove(key);
+         _state.erase(key);
+      }
+
+      /// Look for a key in `_state`. If a value is not found a
+      /// `nullptr` is returned. note that this function will not
+      /// throw if it does not find the key that it's looking for.
+      auto find(const uint64_t& key) -> decltype(&*_state.find(key)) {
+         auto itr = _state.find(key);
+         if (itr != _state.cend()) {
+            return &*itr;
+         }
+         else {
+            return nullptr;
+         }
+      }
+
+      /// Look for a key in `_state` by calling the function `find` on
+      /// its behalf. If a value is not found this function will throw
+      /// an error.
+      auto get(const uint64_t& key) -> decltype(*find(key)) {
+         auto ptr = find(key);
+         if(!ptr) {
+            std::stringstream ss;
+            ss << "Key not found!\n"
+               << "(" << boost::core::demangle(typeid(key).name()) << "): " << key;
+            BOOST_THROW_EXCEPTION(std::out_of_range{ss.str().c_str()});
+         }
+         else {
+            return *ptr;
+         }
+      }
+
+      /// Undo any combination of the following actions done to the
+      /// `_state`: adding key/value pairs, modifying key/value pairs,
+      /// and removing keys (key/value pairs).
+      void undo() {
+         if(!_enabled()) {
+            return;
+         }
+
+         _undo_new_keys(std::move(_stack.back()));
+         _undo_modified_values(std::move(_stack.back()));
+         _undo_removed_values(std::move(_stack.back()));
```

Comment:
See @swatanabe-b1's comment about not moving more than once: https://github.com/EOSIO/chainbase/pull/51/files#commitcomment-34481085

Reply:
Resolved in [28f37d5c88015e37e687a4d79b33de440345e377](https://github.com/EOSIO/chainbase/blob/28f37d5c88015e37e687a4d79b33de440345e377/include/chainbase/chainrocks.hpp#L151-L167).